spark/dl/src/main/scala/org/apache/spark/ml/MLParams.scala

```diff
@@ -0,0 +1,368 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.ml
+
+import com.intel.analytics.bigdl.optim._
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.tensor.{DoubleType, FloatType}
+import org.apache.commons.lang3.SerializationUtils
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.ml.OptimMethodParam._
+import org.apache.spark.ml.param.{Param, Params}
+import org.apache.spark.ml.util.{DefaultParamsReader, DefaultParamsWriter}
+import org.json4s.JsonDSL._
+import org.json4s._
+import org.json4s.jackson.JsonMethods._
+
+import scala.reflect.ClassTag
+
+/**
+ * :: DeveloperApi ::
+ * A param wrapper for [[com.intel.analytics.bigdl.optim.OptimMethod]]
+ */
+@DeveloperApi
+class OptimMethodParam(
+    parent: Params,
+    name: String,
+    doc: String,
+    isValid: OptimMethod[_] => Boolean)
+  extends Param[OptimMethod[_]](parent, name, doc, isValid) {
+
+  def this(parent: Params, name: String, doc: String) =
+    this(parent, name, doc, (_: OptimMethod[_]) => true)
+
+  override def jsonDecode(json: String): OptimMethod[_] = {
+    implicit val formats: DefaultFormats.type = DefaultFormats
+    val js = parse(json)
+    (js \ "tensorDataType").extract[String] match {
+      case "TensorDouble" =>
+        decode[Double]((js \ "objectData").asInstanceOf[JObject])
+      case "TensorFloat" =>
+        decode[Float]((js \ "objectData").asInstanceOf[JObject])
+      case tt =>
+        throw new IllegalArgumentException(s"Wrong TensorDataType: $tt")
+    }
+  }
+
+  override def jsonEncode(value: OptimMethod[_]): String = {
+    // infer tensor data type by suffix of ClassName
+    val json = value.getClass.getName match {
```

Comment:
This is a little weird... Should we consider adding class tag to OptimMethod? @hhbyyh Do you have any commets about this?

Reply:
I haven't tried with the json serialization much myself, yet I'm curious if the current implementation can provide backward compatibility. E.g. if we add a new field for SGD() in a new release, will the current implementation with jObject be able to load the optimMethod saved with previous version?

Reply:
@hhbyyh It's a potential problem, maybe we can add decoding part for new field with default, like : __Option(...).getOrElse(<default_values>)__ ?

Reply:
I don't yet have an easy solution for all the customized parameters (like optimMethod, Trigger, Train/validation summary). For now I would suggest excluding those fields from model serialization. That's just my personal opinion.

Please confirm with @yiheng or @qiuxin2012